tests/lib/rules/no-early-return.js

```diff
@@ -0,0 +1,302 @@
+          function bar () {
+            return 1
+          }
+          return { foo }
+        }
+      }
+      </script>
+      `
+    },
+    {
+      filename: 'test.vue',
+      code: `
+      <script>
+      export default {
+        data () {
+          return { foo: true }
+        }
+      }
+      </script>
+      `
+    },
+    {
+      filename: 'test.vue',
+      code: `
+      <script>
+      export default {
+        asyncData () {
+          return { foo: true }
+        }
+      }
+      </script>
+      `
+    },
+    {
+      filename: 'test.vue',
+      code: `
+      <script>
+      export default {
+        asyncData (payload) {
+          return { foo: true }
+        }
+      }
+      </script>
+      `
+    },
+    {
+      filename: 'test.vue',
+      code: `
+      <script>
+      export default {
+        data: () => ({ foo: true })
+      }
+      </script>
+      `
+    },
+    {
+      filename: 'test.vue',
+      code: `
+      <script>
+      export default {
+        asyncData: () => ({ foo: true })
+      }
+      </script>
+      `
+    },
+    {
+      filename: 'test.vue',
+      code: `
+      <script>
+      export default {
+        setup () {
+          const foo = computed(() => {
+            if (s) {
+              return
+            }
+          })
+        }
+      }
+      </script>
+      `
+    }
+  ],
+  invalid: [
+    {
+      filename: 'test.vue',
+      code: `
+      <script>
+      export default {
+        setup () {
+          const foo = ref()
+
+          function bar () {
+            return 1
+          }
+
+          for (const f of [1]) {
+            if (1) {
+              return 2
+            }
+          }
+
+          if (foo) {
+            return
+          }
+
+          return { foo }
+        }
+      }
+      </script>
+      `,
+      errors: [
+        {
+          messageId: 'extraReturnStatement',
+          data: { functionType: 'setup' },
+          line: 13
+        },
+        {
+          messageId: 'extraReturnStatement',
+          data: { functionType: 'setup' },
+          line: 18
+        }
+      ]
+    },
+    {
+      filename: 'test.vue',
+      code: `
+      <script>
+      export default {
+        setup () {
+          const foo = ref()
+
+          if (foo) {
+            return { foo }
+          }
+        }
+      }
+      </script>
+      `,
+      errors: [
+        {
+          messageId: 'extraReturnStatement',
+          data: { functionType: 'setup' },
+          line: 8
+        }
+      ]
+    },
+    {
+      filename: 'test.vue',
+      code: `
+      <script>
+      export default {
+        data () {
+          if (this.bar) {
+            return { foo: false }
+          }
+          return { foo: true }
+        }
+      }
+      </script>
+      `,
+      errors: [
+        {
+          messageId: 'extraReturnStatement',
+          data: { functionType: 'data' },
+          line: 3
+        }
+      ]
+    },
+    {
```

Comment:
Please add one more test case. I'm afraid this will produce too many errors :slightly_frowning_face: 

```suggestion
    {
      filename: 'test.vue',
      code: `
      <script>
      export default {
        setup: () => {
          if (maybe) {
            return { foo: false }
          }
          return { foo: true }
        },
        data() {
          return {
            foo: true
          }
        },
        computed: {
          foo() {
            return true
          }
        }
      }
      </script>
      `,
      errors: [
        {
          messageId: 'extraReturnStatement',
          data: { functionType: 'setup' },
          line: 3
        }
      ]
    },
    {
```

Reply:
I found a false positive in the next test source.

```vue
<script>
export default {
  methods: {
    fn () {
      if (a) {
        return {}; // <- false positive
      }
    }
  },
  setup () {
    const foo = ref()
    return { foo }
  }
}
</script>
```

I think we need to check the scope of ReturnStatement correctly.
The implementation of the return-in-emits-validator rule may be helpful.
https://github.com/vuejs/eslint-plugin-vue/blob/master/lib/rules/return-in-emits-validator.js